### DISCUSSION TITLE
Single-module (and dependencies) runboat

### DISCUSSION THREAD

**Discussion by francesco-ooops** [OCA/odoo-community.org#126] [Ideas]
Picking up from https://github.com/sbidoul/runboat/pull/95 and subsequent announcement [here](https://odoo-community.org/groups/contributors-15/contributors-1221858?mode=thread&date_begin=&date_end=), here's an idea to think about:

the possibility of building a runboat with a specific single module (and its dependencies) is extremely interesting from a functional perspective, as it's something that would help immensely testing a module (both an already merged one or the one in a PR).

### Issue

Testing a module on a runboat for a repo with tens of modules (eg: sale-workflow, purchase-workflow, stock-logistics-warehouse, ...) is like going to a restaurant to try out how a single ingredient - let's say olives - tastes on pizza, but the only pizza in the menu is one with all possible ingredients in the kitchen. It's unlikely you will leave the restaurant with an opinion on olives.

### Disadvantages

This leads to:
- situations where it's almost impossible to perform a basic task in runboat (eg: adding a vendor on a purchase order)
- strong noise when testing
- employing dev and PM time to fix existing modules disrupting runboat features with the only goal of testing another module on runboat
- most of all, it's defeating one of the most important goals of runboat: providing a testing environment to quickly evaluate how a module works.

This last point is especially unhelpful for functionals with less knowledge of repo structure, who only need an instance to test a single module. Specifically, at the last OCA days we've noticed there was quite some confusion on which runboat to use from the current runboat interface 

### Considerations

- in some repos it's important to have a runboat build with all modules installed (eg: verticals, l10n) 
- of course it's important to check if a runboat build with all modules is successful
- having more runboats would increase required resources

### Proposal

In PRs and in readme file for merged modules, provide two links: 

- one for a runboat build with all repo modules installed (current behavior)
- one for a runboat build with only module and its dependencies installed, and possibility to install other modules from the same repo (or maybe even from the whole OCA catalog?)

cc @OCA/oca-consultants

**Comment by vdewulf** [comment]
I personnally agree with @francesco-ooops and love the pizza metaphor, by the way ;)

**Comment by TumbaoJu** [comment]
I totally agree with @francesco-ooops .
I also like the idea of having the module I want to test already installed with its dependencies BUT to have the option to install more modules of the same repository if needed could also be very useful.

**Reply by francesco-ooops** [reply]
Good point! I edited the original message to specify this. 

Ideally, runboat should be an instance where tester can experiment with the set of modules they want to take into consideration.

**Comment by pedrobaeza** [comment]
This is a good idea, but technically very difficult to be executed. The only viable option I see is to have a runboat without installing anything, but with the option to install any module of the repo (and with their dependencies available), so you should install on demand for trying, and having the option to reset it to the initial state for trying other module.

Not sure though if this is possible with PIP system. cc @sbidoul

**Reply by francesco-ooops** [reply]
If no automation is possibile I think we can live with a manual installation of modules, sure! 

What's important is to end up with both the repo-wide build that is needed for code quality, and the module-only build to use for functional testing.

**Reply by TumbaoJu** [reply]
I agree.  For functional testing of a specific module, we could install the modules we need for testing.  It's better than having ALL modules of the repo already installed.

**Comment by sbidoul** [comment]
Sounds complicated to implement.

One possibility is to *always* let runboat install a build with only `base` so you can select between the full and base-only database when connecting.

I currently chose to create the base-only db only when the full install failed, but that is easy to change. It will make the runboat build a little bit slower (because it needs to create the base only database, but it is quite fast so should be okay).

**Comment by sbidoul** [comment]
Done and deployed. https://github.com/sbidoul/runboat/pull/101

**Reply by pedrobaeza** [reply]
So now any runboat instance don't have any module installed?

**Reply by sbidoul** [reply]
No, each runboat instance will have two databases. One with all modules of the repo installed, and one with only `base`. Example [here](http://oca-mis-builder-16-0-92f56807d2ee.runboat.odoo-community.org/)

**Reply by pedrobaeza** [reply]
Ah, ok, that's great and very quick to be done. That deserves an announce in the list.

**Reply by TumbaoJu** [reply]
Youhou!!!

**Reply by loida-vm** [reply]
Thank you @sbidoul will help a lot with functional tests 👏🏼

**Comment by Gelojr** [comment]
Thanks @francesco-ooops @sbidoul Nice job

**Comment by EmilioPascual** [comment]
Amazing! Thank you @sbidoul!

**Comment by francesco-ooops** [comment]
Thanks a lot @sbidoul for the super fast implementation!

Do you think this could be done also in readme files for modules? I mean having the "Runboat" button point to a build with the double implementation.

**Reply by sbidoul** [reply]
So you want runboat to generate the README.rst from the modified readme fragments on a PR, so it can be seen in the Odoo Apps list?

**Reply by francesco-ooops** [reply]
Actually I just checked and the "base-only" build is already accessible from readme on github:

![image](https://github.com/OCA/odoo-community.org/assets/55489873/07c72bbc-a242-4663-ab23-942d02d6b90d)

So you can ignore my request. Maybe we'll come out later with some proposals to improve ux to get functionals to easily test the module they want to try out, but for now thanks a lot for the work you've done! I can see it's already having a relevant impact on how modules are tested.

**Reply by sbidoul** [reply]
Ah I understand now. Thanks a lot for the feedback!

**Comment by francesco-ooops** [comment]
@sbidoul another thing that would be nice to improve the flow: 

- open runboat from pr X
- choose build "all repo modules installed"
- close runboat > go back to PR > reopen runboat

CURRENT BEHAVIOR

"all repo modules installed" build is automatically open

EXPECTED BEHAVIOR

- user can choose again on which runboat build to test

**Reply by sbidoul** [reply]
If you log out you can select the db on the login screen.

**Reply by francesco-ooops** [reply]
Hi @sbidoul I checked again and I can't seem to access the other runboat by logging out: https://recordit.co/Rtwo3Fijaa

**Reply by sbidoul** [reply]
When only the `baseonly` db is present, it means the installation of the full database has failed. You need to look into the init log to understand why.
